zerver/webhooks/slack/view.py

```diff
@@ -1,60 +1,187 @@
+        .get("app_id")
+        .tame(check_none_or(check_string))
+    )
+    return bot_app_id is not None and app_api_id == bot_app_id
+
+
+def convert_mentions(text: str) -> str:
+    tokens = text.split(" ")
+    for iterator in range(len(tokens)):
+        slack_usermention_match = re.search(SLACK_USERMENTION_REGEX, tokens[iterator], re.VERBOSE)
+        slack_channelmention_match = re.search(
+            SLACK_CHANNELMENTION_REGEX, tokens[iterator], re.MULTILINE
+        )
+        if slack_usermention_match:
+            # TODO: Callback to Slack API
+            pass
+        elif slack_channelmention_match:
+            # Convert Slack channel mentions to a mention-like syntax so that
+            # a mention isn't triggered for a Zulip channel with the same name.
+            channel_info: list[str] = slack_channelmention_match.group(0).split("|")
+            channel_name = channel_info[1]
+            tokens[iterator] = f"**#{channel_name}**" if channel_name else "**#[private channel]**"
+    text = " ".join(tokens)
+    return text
+
+
+# This is a modified version of `convert_to_zulip_markdown` in
+# `slack_message_conversion.py`, which cannot be used directly
+# due to differences in the Slack import data and Slack webhook
+# payloads.
+def convert_to_zulip_markdown(text: str) -> str:
+    text = convert_markdown_syntax(text, SLACK_BOLD_REGEX, "**")
+    text = convert_markdown_syntax(text, SLACK_STRIKETHROUGH_REGEX, "~~")
+    text = convert_markdown_syntax(text, SLACK_ITALIC_REGEX, "*")
+
+    # Map Slack's <!everyone>, <!channel> and <!here> mentions to @**all**.
+    # No regex for this as it can be present anywhere in the sentence.
+    text = text.replace("<!everyone>", "@**all**")
+    text = text.replace("<!channel>", "@**all**")
+    text = text.replace("<!here>", "@**all**")
+
+    text = convert_mentions(text)
+    text, _ = convert_link_format(text)
+    text, _ = convert_mailto_format(text)
+    return text
+
+
+def convert_raw_file_data(file_dict: WildValue) -> SlackFileListT:
+    files = [
+        {
+            "file_link": file.get("permalink").tame(check_string),
+            "file_name": file.get("title").tame(check_string),
+        }
+        for file in file_dict
+    ]
+    return files
+
+
+def get_message_body(text: str, sender: str, files: SlackFileListT) -> str:
+    # TODO: Reformat mentions in messages, currently its <@USER_ID>.
+    body = ZULIP_MESSAGE_TEMPLATE.format(sender=sender, text=text)
+    for file in files:
+        body += FILE_LINK_TEMPLATE.format(**file)
+    return body
+
+
+def is_challenge_handshake(payload: WildValue) -> bool:
+    return payload.get("type").tame(check_string) == "url_verification"
+
 
 @webhook_view("Slack", notify_bot_owner_on_invalid_json=False)
 @typed_endpoint
 def api_slack_webhook(
     request: HttpRequest,
     user_profile: UserProfile,
     *,
-    user_name: str,
-    text: str,
-    channel_name: str,
+    payload: JsonBodyPayload[WildValue],
```

Comment:
Can we make this support the old Slack API as well by just leaving the old code as an `api_slack_legacy_webhook` function that we call if the payload format matches the old Slack API rather than the new one?

That's how we usually try to handle third-party API changes like this. The other option is to change the endpoint URL to `slack_events` or something, but it's usually nicer.

Reply:
Oh, yeah, that’s doable. We could create an adapter function that returns a WildValue payload if we receive the old payload. This way, we can just remove the function if we decide to drop support for the old payload later down the line.